client/my-sites/site-settings/jetpack-credentials/credentials-form/index.jsx

```diff
@@ -0,0 +1,261 @@
+								<FormLabel>
+									<div>{ translate( 'Credential Type' ) }</div>
+									<FormSelect
+										name="protocol"
+										value={ get( this.state.form, 'protocol', 'ssh' ) }
+										onChange={ this.handleFieldChange }
+										disabled={ formIsSubmitting }
+									>
+										<option value="ssh">{ translate( 'SSH' ) }</option>
+										<option value="sftp">{ translate( 'SFTP' ) }</option>
+										<option value="ftp">{ translate( 'FTP' ) }</option>
+									</FormSelect>
+								</FormLabel>
+							</td>
+						</tr>
+						<tr>
+							<td className="credentials-form__host-field">
+								<FormLabel>
+									<div>{ translate( 'Server Address' ) }</div>
+									<FormTextInput
+										name="host"
+										placeholder={ translate( 'yoursite.com' ) }
+										value={ get( this.state.form, 'host', '' ) }
+										onChange={ this.handleFieldChange }
+										disabled={ formIsSubmitting }
+										isError={ !! formErrors.host }
+									/>
+									{ formErrors.host ? (
+										<FormInputValidation isError={ true } text={ formErrors.host } />
+									) : null }
+								</FormLabel>
+							</td>
+							<td className="credentials-form__port-field">
+								<FormLabel>
+									<div>{ translate( 'Port Number' ) }</div>
+									<FormTextInput
+										name="port"
+										placeholder={ translate( '22' ) }
+										value={ get( this.state.form, 'port', '' ) }
+										onChange={ this.handleFieldChange }
+										disabled={ formIsSubmitting }
+										isError={ !! formErrors.port }
+									/>
+									{ formErrors.port && (
+										<FormInputValidation isError={ true } text={ formErrors.port } />
+									) }
+								</FormLabel>
+							</td>
+						</tr>
+						<tr>
+							<td colSpan="2" className="credentials-form__user-field">
+								<FormLabel>
+									<div>{ translate( 'Username' ) }</div>
+									<FormTextInput
+										name="user"
+										placeholder={ translate( 'username' ) }
+										value={ get( this.state.form, 'user', '' ) }
+										onChange={ this.handleFieldChange }
+										disabled={ formIsSubmitting }
+										isError={ !! formErrors.user }
+									/>
+									{ formErrors.user && (
+										<FormInputValidation isError={ true } text={ formErrors.user } />
+									) }
+								</FormLabel>
+							</td>
+						</tr>
+						<tr>
+							<td colSpan="2" className="credentials-form__pass-field">
+								<FormLabel>
+									<div>{ translate( 'Password' ) }</div>
+									<FormTextInput
+										name="pass"
+										placeholder={ translate( 'password' ) }
+										value={ get( this.state.form, 'pass', '' ) }
+										onChange={ this.handleFieldChange }
+										disabled={ formIsSubmitting }
+										isError={ !! formErrors.pass }
+									/>
+									{ formErrors.pass && (
+										<FormInputValidation isError={ true } text={ formErrors.pass } />
+									) }
+								</FormLabel>
+							</td>
+						</tr>
```

Comment:
After all this refactoring it may be more clear  how we can abstract some of this JSX…

what structure is shared among all the fields?
what information is different for each one?
how much of that different data is actually the same thing repeated multiple times (like a name)?

if we can get it to a few key properties then we can probably iterate of the descriptive data structure in a `map()`

```js
{ fields.map( ( { name, placeholder } ) => (
	<tr>
		…
		<FormTextInput { ...{
			name,
			placeholder: placeholder(), // needs to be like () => translate( … ) so translate gets called
			value: get( this.state.form, name, '' ),
			onChange={ this.handleFieldChange },
			isError: !! formErrors[ name ],
		} } />
	</tr>
) }
```

Reply:
I tried this in the latest commit - it doesn't save a ton of code, but is nicer.